File: langs/python/src/dotlyte/typed.py
"""Typed configuration API for DOTLYTE.

Provides ``create_typed_config()`` with schema-driven validation, type
coercion, server/client boundary enforcement, and an immutable result.

Example::

    from dotlyte.typed import create_typed_config

    env = create_typed_config({
        "DATABASE_URL": {"type": "string", "format": "url", "required": True},
        "PORT": {"type": "integer", "default": 3000},
        "DEBUG": {"type": "boolean", "default": False},
        "LOG_LEVEL": {
            "type": "string",
            "enum": ["debug", "info", "warn", "error"],
            "default": "info",
        },
    })

    env["PORT"]       # 3000  (int)
    env["DEBUG"]      # False (bool)
"""

from __future__ import annotations

import os
import re
import types
from typing import Any, Callable, Literal, TypedDict

from dotlyte.boundaries import create_boundary_proxy
from dotlyte.errors import DotlyteError
from dotlyte.loader import load

# ──── Schema Descriptor Types ────

TypeName = Literal["string", "integer", "number", "boolean", "url"]


class FieldDescriptor(TypedDict, total=False):
    """A single field descriptor in the typed config schema."""

    type: TypeName
    required: bool
    default: Any
    enum: list[Any]
    min: float | int
    max: float | int
    sensitive: bool
    doc: str


# ──── Type-coercion helpers ────

_TRUE_VALUES = frozenset({"true", "yes", "1", "on"})
_FALSE_VALUES = frozenset({"false", "no", "0", "off"})
_URL_PATTERN = re.compile(r"^https?://\S+")


def _validate_field(key: str, raw: Any, descriptor: FieldDescriptor) -> Any:
    """Validate and coerce a single value against its field descriptor.

    Args:
        key: The environment variable / config key name.
        raw: The raw value from the merged config.
        descriptor: Schema descriptor for this field.

    Returns:
        The validated and coerced value.

    Raises:
        DotlyteError: On validation failures.

    """
    is_required = descriptor.get("required", True)
    has_default = "default" in descriptor
    field_type: TypeName = descriptor.get("type", "string")

    # Apply default
    value = raw
    if (value is None or value == "") and has_default:
        value = descriptor["default"]

    # Check required
    if (value is None or value == "") and is_required:
        doc_hint = f" ({descriptor['doc']})" if descriptor.get("doc") else ""
        raise DotlyteError(
            f"Missing required environment variable '{key}'.{doc_hint} "
            f"Set it in your .env file, config file, or as an environment variable.",
            key=key,
            code="MISSING_REQUIRED_KEY",
        )

    # Optional and absent
    if value is None or value == "":
        return None

    # ── Type coercion ──
    str_value = str(value)

    if field_type == "boolean":
        if isinstance(value, bool):
            pass  # already correct
        else:
            lower = str_value.lower()
            if lower in _TRUE_VALUES:
                value = True
            elif lower in _FALSE_VALUES:
                value = False
            else:
                raise DotlyteError(
                    f"Environment variable '{key}' expected boolean, got '{str_value}'.",
                    key=key,
                    code="VALIDATION_ERROR",
                )

    elif field_type in ("integer", "number"):
        if isinstance(value, (int, float)) and not isinstance(value, bool):
            if field_type == "integer" and not isinstance(value, int):
                raise DotlyteError(
                    f"Environment variable '{key}' expected integer, got '{str_value}'.",
                    key=key,
                    code="VALIDATION_ERROR",
                )
        else:
            try:
                num = float(str_value)
            except ValueError:
                raise DotlyteError(
                    f"Environment variable '{key}' expected number, got '{str_value}'.",
                    key=key,
                    code="VALIDATION_ERROR",
                )
            if field_type == "integer":
                if not num.is_integer():
                    raise DotlyteError(
                        f"Environment variable '{key}' expected integer, got '{str_value}'.",
                        key=key,
                        code="VALIDATION_ERROR",
                    )
                value = int(num)
            else:
                value = num

    elif field_type in ("string", "url"):
        value = str_value

    # ── Format validation ──
    fmt = descriptor.get("format") or ("url" if field_type == "url" else None)
    if fmt == "url":
        if not _URL_PATTERN.match(str(value)):
            raise DotlyteError(
                f"Environment variable '{key}' is not a valid URL: '{value}'.",
                key=key,
                code="VALIDATION_ERROR",
            )

    # ── Enum validation ──
    enum_vals = descriptor.get("enum")
    if enum_vals is not None and value not in enum_vals:
        joined = ", ".join(str(e) for e in enum_vals)
        raise DotlyteError(
            f"Environment variable '{key}' must be one of [{joined}], got '{value}'.",
            key=key,
            code="VALIDATION_ERROR",
        )

    # ── Min / Max validation ──
    min_val = descriptor.get("min")
    max_val = descriptor.get("max")

    if isinstance(value, (int, float)) and not isinstance(value, bool):
        if min_val is not None and value < min_val:
            raise DotlyteError(
                f"Environment variable '{key}' value {value} is below minimum {min_val}.",
                key=key,
                code="VALIDATION_ERROR",
            )
        if max_val is not None and value > max_val:
            raise DotlyteError(
                f"Environment variable '{key}' value {value} exceeds maximum {max_val}.",
                key=key,
                code="VALIDATION_ERROR",
            )

    if isinstance(value, str):
        if min_val is not None and len(value) < min_val:
            raise DotlyteError(
                f"Environment variable '{key}' length {len(value)} "
                f"is below minimum {min_val}.",
                key=key,
                code="VALIDATION_ERROR",
            )
        if max_val is not None and len(value) > max_val:
            raise DotlyteError(
                f"Environment variable '{key}' length {len(value)} "
                f"exceeds maximum {max_val}.",
                key=key,
                code="VALIDATION_ERROR",
            )

    return value


# ──── Sectioned Schema Support ────


class SectionedSchema(TypedDict, total=False):
    """Schema with server/client/shared sections."""

    server: dict[str, FieldDescriptor]
    client: dict[str, FieldDescriptor]
    shared: dict[str, FieldDescriptor]
    client_prefix: str


def _is_sectioned(schema: dict[str, Any]) -> bool:
    """Return True if schema uses server/client sections."""
    return "server" in schema and "client" in schema


# ──── Main API ────


def create_typed_config(
    schema: dict[str, Any],
    *,
    skip_validation: bool = False,
    on_secret_access: Callable[[str, str], None] | None = None,
    **load_options: Any,
) -> types.MappingProxyType[str, Any]:
    """Create a validated, typed, immutable configuration object.

    Validates at import-time (fail fast on app startup). Returns a frozen
    ``MappingProxyType`` so the config cannot be mutated.

    Supports two schema formats:

    * **Flat schema** — ``{"PORT": {"type": "integer", "default": 3000}}``
    * **Sectioned schema** — ``{"server": {...}, "client": {...}, "shared": {...}}``

    Args:
        schema: A flat or sectioned ``FieldDescriptor`` schema.
        skip_validation: Skip validation (useful in tests).
        on_secret_access: Audit callback when a sensitive key is read.
        **load_options: Forwarded to ``load()``.

    Returns:
        An immutable mapping of validated config values.

    Raises:
        DotlyteError: On validation failures.

    Example::

        env = create_typed_config({
            "PORT": {"type": "integer", "default": 3000},
            "DEBUG": {"type": "boolean", "default": False},
        })
        env["PORT"]  # 3000

    """
    if _is_sectioned(schema):
        return _create_sectioned_config(
            schema,
            skip_validation=skip_validation,
            on_secret_access=on_secret_access,
            **load_options,
        )
    return _create_flat_config(
        schema,
        skip_validation=skip_validation,
        **load_options,
    )


def _create_flat_config(
    schema: dict[str, FieldDescriptor],
    *,
    skip_validation: bool = False,
    **load_options: Any,
) -> types.MappingProxyType[str, Any]:
    """Handle flat (non-sectioned) schemas."""
    config = load(**load_options)
    raw = config.to_dict()
    env_vars = dict(os.environ)

    result: dict[str, Any] = {}
    errors: list[str] = []

    for key, descriptor in schema.items():
        try:
            raw_value = raw.get(key.lower()) or raw.get(key) or env_vars.get(key)
            result[key] = _validate_field(key, raw_value, descriptor)
        except DotlyteError as exc:
            if skip_validation:
                result[key] = (
                    raw.get(key.lower()) or raw.get(key) or env_vars.get(key)
                )
            else:
                errors.append(str(exc))

    if errors:
        raise DotlyteError(
            f"Typed config validation failed ({len(errors)} error(s)):\n"
            + "\n".join(f"  • {m}" for m in errors),
            code="VALIDATION_ERROR",
        )

    return types.MappingProxyType(result)


def _create_sectioned_config(
    schema: dict[str, Any],
    *,
    skip_validation: bool = False,
    on_secret_access: Callable[[str, str], None] | None = None,
    **load_options: Any,
) -> types.MappingProxyType[str, Any]:
    """Handle sectioned (server/client/shared) schemas."""
    server_schema: dict[str, FieldDescriptor] = schema.get("server", {})
    client_schema: dict[str, FieldDescriptor] = schema.get("client", {})
    shared_schema: dict[str, FieldDescriptor] = schema.get("shared", {})
    client_prefix: str = schema.get("client_prefix", "NEXT_PUBLIC_")

    # Validate that client keys start with the prefix
    for key in client_schema:
        if not key.startswith(client_prefix):
            raise DotlyteError(
                f"Client environment variable '{key}' must start with "
                f"'{client_prefix}'. Move it to the 'server' section or rename it.",
                key=key,
                code="VALIDATION_ERROR",
            )

    # Merge all into one flat schema for validation
    all_schema = {**server_schema, **client_schema, **shared_schema}
    flat = _create_flat_config(all_schema, skip_validation=skip_validation, **load_options)

    # Build key sets
    server_keys = set(server_schema.keys())
    client_keys = set(client_schema.keys())
    shared_keys = set(shared_schema.keys())

    proxy = create_boundary_proxy(
        dict(flat),
        server_keys=server_keys,
        client_keys=client_keys,
        shared_keys=shared_keys,
        on_secret_access=on_secret_access,
    )

    # Build dict from proxy (iter yields keys, getitem yields values)
    result_dict = {k: proxy[k] for k in proxy}
    return types.MappingProxyType(result_dict)

File: langs/rust/src/typed.rs
//! Typed configuration with schema validation — DOTLYTE v0.1.2.
//!
//! Define field schemas with types, constraints, and defaults, then validate
//! environment variables and loaded config against that schema.

use std::collections::HashMap;
use std::fmt;

use serde_json::Value;

use crate::coercion::coerce_str;
use crate::errors::DotlyteError;

/// Field types supported by the typed config system.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum FieldType {
    /// A UTF-8 string value.
    String,
    /// An integer value (i64).
    Integer,
    /// A floating-point number (f64).
    Number,
    /// A boolean value.
    Boolean,
    /// A URL string (must contain `://`).
    Url,
}

impl fmt::Display for FieldType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            FieldType::String => write!(f, "string"),
            FieldType::Integer => write!(f, "integer"),
            FieldType::Number => write!(f, "number"),
            FieldType::Boolean => write!(f, "boolean"),
            FieldType::Url => write!(f, "url"),
        }
    }
}

/// A single field descriptor defining type, constraints, and metadata.
#[derive(Debug, Clone)]
pub struct FieldDescriptor {
    /// The expected field type.
    pub field_type: FieldType,
    /// Whether the field is required (default: `true`).
    pub required: bool,
    /// Fallback value when the field is missing.
    pub default_value: Option<Value>,
    /// Restrict values to this set.
    pub allowed_values: Option<Vec<Value>>,
    /// Minimum value (numbers) or minimum length (strings).
    pub min: Option<f64>,
    /// Maximum value (numbers) or maximum length (strings).
    pub max: Option<f64>,
    /// Marks the field as containing secret data.
    pub sensitive: bool,
    /// Human-readable description of the field.
    pub doc: Option<String>,
}

impl Default for FieldDescriptor {
    fn default() -> Self {
        Self {
            field_type: FieldType::String,
            required: true,
            default_value: None,
            allowed_values: None,
            min: None,
            max: None,
            sensitive: false,
            doc: None,
        }
    }
}

/// Options controlling `create_typed_config` behaviour.
#[derive(Debug, Clone, Default)]
pub struct TypedConfigOptions {
    /// When `true`, skip validation and return raw values.
    pub skip_validation: bool,
}

/// Create a validated, typed configuration map from environment variables and
/// loaded config, validated against `schema`.
///
/// Values are read from `std::env::var` (case-sensitive key first, then
/// upper-cased), then coerced to the declared [`FieldType`]. Constraints
/// (`required`, `allowed_values`, `min`, `max`) are checked unless
/// [`TypedConfigOptions::skip_validation`] is set.
///
/// # Errors
///
/// Returns [`DotlyteError::ValidationError`] when one or more fields fail
/// validation.
pub fn create_typed_config(
    schema: &HashMap<String, FieldDescriptor>,
    options: Option<TypedConfigOptions>,
) -> crate::Result<HashMap<String, Value>> {
    let opts = options.unwrap_or_default();

    let mut result = HashMap::with_capacity(schema.len());
    let mut errors: Vec<crate::errors::SchemaViolation> = Vec::new();

    for (key, desc) in schema {
        let raw = lookup_env_value(key);

        match validate_typed_field(key, raw.as_ref(), desc) {
            Ok(Some(val)) => {
                result.insert(key.clone(), val);
            }
            Ok(None) => {
                // optional and absent — skip
            }
            Err(e) => {
                if opts.skip_validation {
                    if let Some(r) = raw {
                        result.insert(key.clone(), r);
                    }
                } else {
                    errors.push(crate::errors::SchemaViolation {
                        key: key.clone(),
                        message: e.to_string(),
                        rule: "typed".to_string(),
                    });
                }
            }
        }
    }

    if !errors.is_empty() {
        return Err(DotlyteError::ValidationError { violations: errors });
    }

    Ok(result)
}

// ── Internal helpers ────────────────────────────────────────────

/// Look up a value from environment variables.
fn lookup_env_value(key: &str) -> Option<Value> {
    // Try exact key first
    if let Ok(v) = std::env::var(key) {
        if !v.is_empty() {
            return Some(coerce_str(&v));
        }
    }
    // Try upper-cased key
    let upper = key.to_uppercase();
    if upper != key {
        if let Ok(v) = std::env::var(&upper) {
            if !v.is_empty() {
                return Some(coerce_str(&v));
            }
        }
    }
    // Try lower-cased key
    let lower = key.to_lowercase();
    if lower != key && lower != upper {
        if let Ok(v) = std::env::var(&lower) {
            if !v.is_empty() {
                return Some(coerce_str(&v));
            }
        }
    }
    None
}

/// Validate and coerce a single field value.
fn validate_typed_field(
    key: &str,
    raw: Option<&Value>,
    desc: &FieldDescriptor,
) -> Result<Option<Value>, DotlyteError> {
    // Determine effective value
    let value = match raw {
        Some(v) if !is_empty_value(v) => v.clone(),
        _ => {
            if let Some(ref def) = desc.default_value {
                def.clone()
            } else if desc.required {
                let mut msg = format!("missing required config key '{key}'");
                if let Some(ref doc) = desc.doc {
                    msg.push_str(&format!(" ({doc})"));
                }
                msg.push_str(
                    ". Set it in your .env file, config file, or as an environment variable.",
                );
                return Err(DotlyteError::MissingKey {
                    key: key.to_string(),
                    sources_checked: vec!["env".to_string()],
                });
            } else {
                return Ok(None);
            }
        }
    };

    // Coerce to target type
    let coerced = coerce_to_type(key, &value, &desc.field_type)?;

    // Enum validation
    if let Some(ref allowed) = desc.allowed_values {
        let coerced_str = value_to_string(&coerced);
        let found = allowed.iter().any(|a| value_to_string(a) == coerced_str);
        if !found {
            return Err(DotlyteError::ValidationError {
                violations: vec![crate::errors::SchemaViolation {
                    key: key.to_string(),
                    message: format!(
                        "value {coerced} not in allowed values: {allowed:?}"
                    ),
                    rule: "enum".to_string(),
                }],
            });
        }
    }

    // Min/Max for numbers
    if let Some(num) = coerced.as_f64() {
        if let Some(min) = desc.min {
            if num < min {
                return Err(DotlyteError::ValidationError {
                    violations: vec![crate::errors::SchemaViolation {
                        key: key.to_string(),
                        message: format!("value {num} is below minimum {min}"),
                        rule: "min".to_string(),
                    }],
                });
            }
        }
        if let Some(max) = desc.max {
            if num > max {
                return Err(DotlyteError::ValidationError {
                    violations: vec![crate::errors::SchemaViolation {
                        key: key.to_string(),
                        message: format!("value {num} exceeds maximum {max}"),
                        rule: "max".to_string(),
                    }],
                });
            }
        }
    }

    // Min/Max for strings (length)
    if let Some(s) = coerced.as_str() {
        let len = s.len() as f64;
        if let Some(min) = desc.min {
            if len < min {
                return Err(DotlyteError::ValidationError {
                    violations: vec![crate::errors::SchemaViolation {
                        key: key.to_string(),
                        message: format!(
                            "string length {len} is below minimum {min}"
                        ),
                        rule: "min".to_string(),
                    }],
                });
            }
        }
        if let Some(max) = desc.max {
            if len > max {
                return Err(DotlyteError::ValidationError {
                    violations: vec![crate::errors::SchemaViolation {
                        key: key.to_string(),
                        message: format!(
                            "string length {len} exceeds maximum {max}"
                        ),
                        rule: "max".to_string(),
                    }],
                });
            }
        }
    }

    Ok(Some(coerced))
}

/// Coerce a `Value` to the target [`FieldType`].
fn coerce_to_type(key: &str, value: &Value, field_type: &FieldType) -> crate::Result<Value> {
    let str_value = value_to_string(value);

    match field_type {
        FieldType::Boolean => {
            if let Some(b) = value.as_bool() {
                return Ok(Value::Bool(b));
            }
            match str_value.to_lowercase().as_str() {
                "true" | "yes" | "1" | "on" => Ok(Value::Bool(true)),
                "false" | "no" | "0" | "off" => Ok(Value::Bool(false)),
                _ => Err(DotlyteError::ValidationError {
                    violations: vec![crate::errors::SchemaViolation {
                        key: key.to_string(),
                        message: format!("expected boolean, got '{str_value}'"),
                        rule: "type".to_string(),
                    }],
                }),
            }
        }
        FieldType::Integer => {
            if let Some(i) = value.as_i64() {
                return Ok(Value::Number(i.into()));
            }
            if let Some(f) = value.as_f64() {
                let i = f as i64;
                if (f - i as f64).abs() < f64::EPSILON {
                    return Ok(Value::Number(i.into()));
                }
            }
            match str_value.parse::<i64>() {
                Ok(i) => Ok(Value::Number(i.into())),
                Err(_) => Err(DotlyteError::ValidationError {
                    violations: vec![crate::errors::SchemaViolation {
                        key: key.to_string(),
                        message: format!("expected integer, got '{str_value}'"),
                        rule: "type".to_string(),
                    }],
                }),
            }
        }
        FieldType::Number => {
            if value.as_f64().is_some() {
                return Ok(value.clone());
            }
            if let Some(i) = value.as_i64() {
                if let Some(n) = serde_json::Number::from_f64(i as f64) {
                    return Ok(Value::Number(n));
                }
            }
            match str_value.parse::<f64>() {
                Ok(f) => match serde_json::Number::from_f64(f) {
                    Some(n) => Ok(Value::Number(n)),
                    None => Err(DotlyteError::ValidationError {
                        violations: vec![crate::errors::SchemaViolation {
                            key: key.to_string(),
                            message: format!("expected number, got '{str_value}'"),
                            rule: "type".to_string(),
                        }],
                    }),
                },
                Err(_) => Err(DotlyteError::ValidationError {
                    violations: vec![crate::errors::SchemaViolation {
                        key: key.to_string(),
                        message: format!("expected number, got '{str_value}'"),
                        rule: "type".to_string(),
                    }],
                }),
            }
        }
        FieldType::String => Ok(Value::String(str_value)),
        FieldType::Url => {
            if !str_value.contains("://") {
                return Err(DotlyteError::ValidationError {
                    violations: vec![crate::errors::SchemaViolation {
                        key: key.to_string(),
                        message: format!("'{str_value}' is not a valid URL"),
                        rule: "format".to_string(),
                    }],
                });
            }
            Ok(Value::String(str_value))
        }
    }
}

/// Convert a `Value` to its string representation.
fn value_to_string(v: &Value) -> String {
    match v {
        Value::String(s) => s.clone(),
        Value::Number(n) => n.to_string(),
        Value::Bool(b) => b.to_string(),
        Value::Null => String::new(),
        other => other.to_string(),
    }
}

/// Check whether a value is considered empty.
fn is_empty_value(v: &Value) -> bool {
    match v {
        Value::Null => true,
        Value::String(s) => s.is_empty(),
        _ => false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_field_type_display() {
        assert_eq!(FieldType::String.to_string(), "string");
        assert_eq!(FieldType::Integer.to_string(), "integer");
        assert_eq!(FieldType::Boolean.to_string(), "boolean");
        assert_eq!(FieldType::Url.to_string(), "url");
    }

    #[test]
    fn test_coerce_boolean() {
        let val = Value::String("yes".to_string());
        let result = coerce_to_type("test", &val, &FieldType::Boolean).unwrap();
        assert_eq!(result, Value::Bool(true));

        let val = Value::String("off".to_string());
        let result = coerce_to_type("test", &val, &FieldType::Boolean).unwrap();
        assert_eq!(result, Value::Bool(false));
    }

    #[test]
    fn test_coerce_integer() {
        let val = Value::String("42".to_string());
        let result = coerce_to_type("test", &val, &FieldType::Integer).unwrap();
        assert_eq!(result, Value::Number(42.into()));
    }

    #[test]
    fn test_coerce_number() {
        let val = Value::String("3.14".to_string());
        let result = coerce_to_type("test", &val, &FieldType::Number).unwrap();
        assert_eq!(result.as_f64().unwrap(), 3.14);
    }

    #[test]
    fn test_coerce_url_valid() {
        let val = Value::String("https://example.com".to_string());
        let result = coerce_to_type("test", &val, &FieldType::Url).unwrap();
        assert_eq!(result, Value::String("https://example.com".to_string()));
    }

    #[test]
    fn test_coerce_url_invalid() {
        let val = Value::String("not-a-url".to_string());
        assert!(coerce_to_type("test", &val, &FieldType::Url).is_err());
    }

    #[test]
    fn test_validate_default_applied() {
        let desc = FieldDescriptor {
            field_type: FieldType::Integer,
            required: true,
            default_value: Some(Value::Number(3000.into())),
            ..Default::default()
        };
        let result = validate_typed_field("port", None, &desc).unwrap();
        assert_eq!(result, Some(Value::Number(3000.into())));
    }

    #[test]
    fn test_validate_required_missing() {
        let desc = FieldDescriptor {
            field_type: FieldType::String,
            required: true,
            ..Default::default()
        };
        assert!(validate_typed_field("missing", None, &desc).is_err());
    }

    #[test]
    fn test_validate_optional_missing() {
        let desc = FieldDescriptor {
            field_type: FieldType::String,
            required: false,
            ..Default::default()
        };
        let result = validate_typed_field("opt", None, &desc).unwrap();
        assert_eq!(result, None);
    }

    #[test]
    fn test_validate_enum() {
        let desc = FieldDescriptor {
            field_type: FieldType::String,
            required: true,
            allowed_values: Some(vec![
                Value::String("dev".to_string()),
                Value::String("prod".to_string()),
            ]),
            ..Default::default()
        };
        let val = Value::String("dev".to_string());
        assert!(validate_typed_field("env", Some(&val), &desc).is_ok());

        let val = Value::String("staging".to_string());
        assert!(validate_typed_field("env", Some(&val), &desc).is_err());
    }

    #[test]
    fn test_validate_min_max_number() {
        let desc = FieldDescriptor {
            field_type: FieldType::Integer,
            required: true,
            min: Some(1.0),
            max: Some(65535.0),
            ..Default::default()
        };
        let val = Value::Number(8080.into());
        assert!(validate_typed_field("port", Some(&val), &desc).is_ok());

        let val = Value::Number(0.into());
        assert!(validate_typed_field("port", Some(&val), &desc).is_err());

        let val = Value::Number(70000.into());
        assert!(validate_typed_field("port", Some(&val), &desc).is_err());
    }

    #[test]
    fn test_validate_min_max_string_length() {
        let desc = FieldDescriptor {
            field_type: FieldType::String,
            required: true,
            min: Some(3.0),
            max: Some(10.0),
            ..Default::default()
        };
        let val = Value::String("hello".to_string());
        assert!(validate_typed_field("name", Some(&val), &desc).is_ok());

        let val = Value::String("hi".to_string());
        assert!(validate_typed_field("name", Some(&val), &desc).is_err());
    }
}
